Add `counts` and `add_new_item` methods, without tests

Oxygen588 opened the pull request:
Close  #3 , Close #2 ;
Modified the addItem method to not allow duplicate jobs to be added.
Created a new method that gets the length atomically.

Oxygen588 committed:
Implemented no duplicate when adding items and tests, atomic length.

Implemented no duplicate items based on id (A update would be creating the id based on the data hash, as well as currently within some languages add item method getting the list of item could be done using pipeline)
Implemented tests for each one of the languages for the no duplicate items.
Fixed some issues related to run bash file.
Within Job spawner modified the way shared jobs work so its not within the main list as well as fixed the way the host and port is being set.
Python test fixed the way the host and port is being set.

JOT85 committed:
Remove `tests/go/go` binary

JOT85 committed:
Merge branch 'JavaScript-Implementation' into Atomic-length-method

Oxygen588 committed:
Update WorkQueue.ts

AddAtomicItem method was fixed by using transactions instead of pipeline.
The normal AddItem was added back.

Oxygen588 committed:
Atomic functions fix.

GetQueueLengthsAtomic function as well as the AddAtomicItem function has been created for the atomic way of adding items, and the normal AddItem function has been added back.

Oxygen588 committed:
Update index.ts

Change in name for the atomic tests function call

Oxygen588 committed:
Update run-test.sh

Moved the start of job spawner before all the tests so there is no conflict between the new tests.

Oxygen588 committed:
Atomic lengths Python Fix.

Old Add_Item function has been added back, the atomic add_item is now add_atomic_item and has been fixed to run atomically as well as the get_queue_lengths in now atomic fixed as well.

Oxygen588 committed:
Rust-lang Atomicity functions fix

The add_atomic_item has been fixed now, the add_item function has been added back, The tests for duplicate items has been fixed.

Oxygen588 committed:
Update main.rs

Small update to tests, forgot to uncomment the add of the second item.

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +122,67 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+	return err
+}
+
+func (workQueue *WorkQueue) GetQueueLengthsAtomic(ctx context.Context, db *redis.Client) ([]int64, error) {
+	tx := db.TxPipeline()
+
+	queueLen := tx.LLen(ctx, workQueue.mainQueueKey)
+	processingLen := tx.LLen(ctx, workQueue.processingKey)
+
+	_, err := tx.Exec(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]int64, 2)
+	result[0], _ = queueLen.Result()
+	result[1], _ = processingLen.Result()
+
+	return result, nil
+}
+
+func (workQueue *WorkQueue) GetQueueLengths(ctx context.Context, db *redis.Client) ([]int64, error) {
+	pipe := db.Pipeline()
+	queueLen := pipe.LLen(ctx, workQueue.mainQueueKey)
+	processingLen := pipe.LLen(ctx, workQueue.processingKey)
+	_, err := pipe.Exec(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]int64, 2)
+	result[0], _ = queueLen.Result()
+	result[1], _ = processingLen.Result()
+	return result, err
+}
```
I don't think there's a need for the non-atomic method. If you want the queue lengths non-atomically, then you can call the existing separate methods. If you want them both, I think you'd generally assume it's atomic, so I think `GetQueueLengths` should be the transaction version, and that should be the only version?

Also, on the naming, we currently have:
- `QueueLen`
- `Processing`

It might be a more consistent name to call it `Lengths`? Perhaps just `Len` but that might be unclear? What's your opinion?

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +122,67 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
 	return err
 }
 
+func (workQueue *WorkQueue) AddAtomicItem(ctx context.Context, db *redis.Client, item Item) error {
+	tx := db.TxPipeline()
+
+	// Add LRANGE commands to the transaction
+	tx.LRange(ctx, workQueue.mainQueueKey, 0, -1)
+	tx.LRange(ctx, workQueue.processingKey, 0, -1)
+
+	// Execute the transaction
+	results, err := tx.Exec(ctx)
+	if err != nil {
+		return err
+	}
+
+	mainItems := results[0].(*redis.StringSliceCmd).Val()
+	processingItems := results[1].(*redis.StringSliceCmd).Val()
+	itemID := item.ID
+
+	if sliceContainsString(mainItems, itemID) || sliceContainsString(processingItems, itemID) {
+		// Same item tried being added twice.
+		return nil
+	}
+
+	pipeline := db.TxPipeline()
+	workQueue.AddItemToPipeline(ctx, pipeline, item)
+	_, err = pipeline.Exec(ctx)
+	return err
+}
+
```
What happens if, in between the first transaction (on line 133) and adding the item (on line 149), another thread comes and adds the item?

Also, the point of this method is to add an item only if it's not a duplicate. I think the name of the method should make that obvious.

JOT85 commented on go/WorkQueue.go:
```diff
@@ -68,6 +68,8 @@ import (
 	"context"
 	"time"
 
+	//"fmt"
+
```
This can be deleted :)

Oxygen588 replied:
Yep!

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -47,6 +47,35 @@ export class WorkQueue {
+   * @param item The item that will be executed using the method addItemToPipeline.
+   */
+  async addAtomicItem(db: Redis, item: Item): Promise<boolean | null> {
+    return new Promise<boolean | null>((resolve, reject) => {
+      db.multi()
+        .lrange(this.mainQueueKey, 0, -1)
+        .lrange(this.processingKey, 0, -1)
+        .exec(async (err, results) => {
+          if (
+            results &&
+            ((results[0][1] as string[]).includes(item.id) ||
+              (results[1][1] as string[]).includes(item.id))
+          ) {
+            resolve(null);
+          } else {
+            const pipeline = db.pipeline();
+            this.addItemToPipeline(pipeline, item);
+            await pipeline.exec();
+            resolve(true);
+          }
+        });
+    });
```
Can we use `await` here on the first transaction to neaten things up?

Also see the comment about insertion between the transactions from the Go code.

Oxygen588 committed:
Update WorkQueue.go

Removed non Atomic way of getting queue length and renamed the GetQueueLengthsAtomic to GetQueueLengths

Oxygen588 committed:
Update WorkQueue.go

Watch atomicity implemented within go-lang.

Oxygen588 committed:
Update WorkQueue.go

AddAtomicItem Fix!
Added warning message.

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +122,77 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
 	return err
 }
 
+// **************-This method can be used only when database is locked-**************
```
Write a proper doc comment please!

You can see the other comments for examples, and consider how they'll look on https://pkg.go.dev/github.com/mevitae/redis-work-queue/go.

https://tip.golang.org/doc/comment

In the summary you can also explain the function (and feel free to think of a better name).

Oxygen588 replied:
will do

Oxygen588 committed:
Atomicity fully fixed + tests fixed.

The atomic function fully works and tests have been modified to support locking!

Oxygen588 committed:
Update WorkQueue.go

Small fix.

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +122,77 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+		return nil
+	}
+
+	err := db.Watch(ctx, txf, workQueue.mainQueueKey, workQueue.processingKey)
+	//fmt.Println(err)
+	for err == redis.TxFailedErr {
+		if err == nil {
+			// Success.
+			return nil
+		} else if err == redis.TxFailedErr {
+			err = db.Watch(ctx, txf, workQueue.mainQueueKey, workQueue.processingKey)
+		}
+	}
+	return nil
```
This could be much more simple:

```go
for {
    err := db.Watch(ctx, txf, workQueue.mainQueueKey, workQueue.processingKey)
    // Retry on transaction failure, return on anything else.
    if err != redis.TxFailedErr {
        return err
    }
}
```

Oxygen588 replied:
<img width="990" alt="image" src="https://github.com/MeVitae/redis-work-queue/assets/69449240/b0b72158-2e79-4ca7-8122-54499f01ee4c">

JOT85 replied:
The code for calling `Watch` is still duplicated! It just makes it more prone to error in the future when someone comes to change things.

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +122,77 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+	return nil
+}
+
+func (workQueue *WorkQueue) GetQueueLengths(ctx context.Context, db *redis.Client) ([]int64, error) {
```
There's no need to return a slice here! It's much more idiomatic to return two ints.

For example:

```go
func (workQueue *WorkQueue) Lengths(ctx context.Context, db *redis.Client) (int64, int64, error)
```

You can also annotate the return values to make it more readable:

```go
func (workQueue *WorkQueue) Lengths(ctx context.Context, db *redis.Client) (queueLen, processing int64, error)
```

JOT85 replied:
Also, if you ever are returning a fixed length slice, return an array instead (i.e. `[2]int64`)

JOT85 commented:
Sorry I didn't have your latest commit when writing that last review. I'll update my comments.

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +123,85 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+	txf := func(tx *redis.Tx) error {
+		pipe := db.Pipeline()
+
+		mainQueueLenCmd := pipe.LLen(ctx, workQueue.mainQueueKey)
+		processingQueueLenCmd := pipe.LLen(ctx, workQueue.processingKey)
```
How come we need to get the lengths?

Oxygen588 replied:
[in ](https://pkg.go.dev/github.com/go-redis/redis/v8#Client.LPos)
https://pkg.go.dev/github.com/go-redis/redis/v8#LPosArgs

JOT85 replied:
It only uses the max length if it isn't 0: https://github.com/redis/go-redis/blob/v8.11.5/commands.go#L1422 (the library has poor documentation).

Therefore setting it to 0 won't pass the argument, and therefore the whole list will be checked (https://redis.io/commands/lpos/).

In Go, we can just not pass the parameter when creating the struct and it'll default to 0. Similarly, you don't need to specify rank, either.

Oxygen588 committed:
Update WorkQueue.ts

TS addAtomicItem fix using watch.

Oxygen588 committed:
AddAtomicItem Rename + Better Doc Comment

'AddAtomicItem Rename' to 'AddItemAtomically'.
Re-written the comment for AddItemAtomically.

Oxygen588 committed:
Update WorkQueue.go

Better Watch for loop.

Oxygen588 committed:
AddItemAtomically, Lengths small changes

AddItemAtomically: removed unnecessary pipe.
Improved the return of the lengths function.

Oxygen588 committed:
Update WorkQueue.ts

Removed the limit of tries until it stops trying.

Oxygen588 committed:
Fixed add_item_atomically.

The new fix implements the redis watch. This should make it fully atomic.

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +121,72 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+		return 0, 0, err
+	}
+
+	return queueLen, processingLen, err
```
No need to specify the values here. You can just use `return`. You specified them in the signature :)

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +121,72 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+func (workQueue *WorkQueue) Lengths(ctx context.Context, db *redis.Client) (queueLen, processingLen int64, err error) {
+	tx := db.TxPipeline()
+
+	queueLen, err = tx.LLen(ctx, workQueue.mainQueueKey).Result()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	processingLen, err = tx.LLen(ctx, workQueue.processingKey).Result()
+	if err != nil {
+		return 0, 0, err
+	}
```
How does `Result()` work before calling `Exec` on the transaction?

JOT85 replied:
That should have been caught in tests!

JOT85 commented on rust/src/lib.rs:
```diff
@@ -280,12 +280,43 @@ impl WorkQueue {
+    
+        pipe
+            .atomic()
+            .cmd("LRANGE").arg(&main_queue_key).arg(0).arg(-1)
+            .cmd("LRANGE").arg(&processing_key).arg(0).arg(-1);
```
You can use the `lrange` method here!

JOT85 commented on rust/src/lib.rs:
```diff
@@ -280,12 +280,43 @@ impl WorkQueue {
     /// Add an item to the work queue.
     ///
     /// This creates a pipeline and executes it on the database.
```
Update the comment!

JOT85 commented on rust/src/lib.rs:
```diff
@@ -303,6 +334,20 @@ impl WorkQueue {
+    pub async fn get_queue_lengths<'a, C: AsyncCommands>(
+        &'a self,
+        db: &mut C,
+    ) -> RedisResult<Vec<i64>> {
```
Definitely don't return a `Vec`.

A `Vec` is a heap allocated array + capacity. So a `Vec` itself has 3 ints, the base pointer, the length and the capacity. Further to that, it performs a heap allocation of the data.

If we just pass a tuple (`(i64, i64)`) then its just two ints big. Smaller than the `Vec` pointer itself and no heap allocation!!!

Finally, the lengths aren't signed, so this should be `u64` not `i64`.

JOT85 replied:
@Oxygen588 could you please resolve this?

You just need to change the return type to `(u64, u64)` (or `u32`) and then return the result as a tuple. Definitely not a Vec!

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -47,6 +47,64 @@ export class WorkQueue {
+          this.addItemToPipeline(pipeEx, item);
+          await pipeEx.exec();
+          return true;
+        } catch (error) {
+          transactionError = true;
```
This catches any error, just just a transaction failure!

JOT85 replied:
Like in Go, we should only retry if the error is a failed transaction.

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -47,6 +47,64 @@ export class WorkQueue {
+            return false; // Item already exists in either mainQueue or processingKey
+          }
+  
+          const pipeEx = db.pipeline();
+          this.addItemToPipeline(pipeEx, item);
+          await pipeEx.exec();
```
For `WATCH` to work, this has to wrap the command in `MULTI ... EXEC`. I don't think it does?

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -47,6 +47,64 @@ export class WorkQueue {
+    return await result(); // Invoke the async function to get the result
+  }
+  
+  
+  
+  
+  
+  
+
```
No need for this blank space!

JOT85 commented on python/redis_work_queue/workqueue.py:
```diff
@@ -37,6 +37,34 @@ def add_item(self, db: Redis, item: Item) -> None:
+            try:
+                pipeline.watch(self._main_queue_key, self._processing_key)
+
+                main_pos = pipeline.lpos(self._main_queue_key, item.id())
+                processing_pos = pipeline.lpos(self._processing_key, item.id())
+                if main_pos is not None or processing_pos is not None:
+                    break  
```
These aren't actually executed!

JOT85 commented on python/redis_work_queue/workqueue.py:
```diff
@@ -37,6 +37,34 @@ def add_item(self, db: Redis, item: Item) -> None:
+
+        pipeline.unwatch()
+
+        
+
+
```
No need for this blank space.

JOT85 replied:
No need for this blank space.

Oxygen588 committed:
Unexecuted pipe fix Ts, necessary fixes done within addAtomicItem Ts, fixed return at Lengths function Go

Unexecuted pipe fix.
addAtomicItem Pipes fixed and executed properly as well as necessary checks for transaction errors or generic errors has been fixed

Oxygen588 committed:
Update workqueue.py

Fixed unexecuted pipe and properly checked the pipe for items already in list.

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -47,6 +48,61 @@ export class WorkQueue {
+            }
+          });
+        } catch (error) {
+          transactionError = false;
```
I don't think there's a need for this catch?

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -47,6 +48,61 @@ export class WorkQueue {
+   */
+
+  async addAtomicItem(db: Redis, item: Item): Promise<boolean | null> {
+    const result = async (): Promise<boolean> => {
```
Do we need to wrap this body in `result`? I think it's only called once, right after it's declared?

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +120,75 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+	queueLenPipe := tx.LLen(ctx, workQueue.mainQueueKey)
+	processingLenPipe := tx.LLen(ctx, workQueue.processingKey)
+
+	_, err = tx.Exec(ctx)
```
This `err` is never checked!

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +120,75 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+		return 0, 0, err
+	}
+	processingLen, err = processingLenPipe.Result()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	if err != nil {
+		return 0, 0, err
+	}
```
Neither if these are necessary? The error is returned in the final return statement.

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -47,6 +48,61 @@ export class WorkQueue {
+  
+          const pipeEx = db.multi();
+          this.addItemToPipeline(pipeEx, item);
+          await pipeEx.exec((err, results) => {
+            if (results && results[0][1] === "OK") {
+              success = true;
+              transactionError = false;
+              return transactionError;
+            }
```
Do we need to use the callback here? Can we get the result from the promise?

Oxygen588 committed:
Update WorkQueue.go

Removed unnecessary error checks, adding the necessary  error check for the pipeline execution.

Oxygen588 committed:
Update WorkQueue.ts

Unwrapped result as well as removed the callback and modified it to a normal varialble.

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,6 +120,68 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+		return 0, 0, err
+	}
+
+	queueLen, err = queueLenPipe.Result()
```
Now this `err` is never checked!! Attention to detail :D

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -47,6 +48,54 @@ export class WorkQueue {
+  
+
+          if (pipeResult && (pipeResult[0][1] !== null || pipeResult[1][1] !== null)) {
+            transactionError = false; // Item already exists in either mainQueue or processingKey
```
No need to set this, you're about to return!

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -47,6 +48,54 @@ export class WorkQueue {
+          this.addItemToPipeline(pipeEx, item);
+          let results = await pipeEx.exec()
+          if (results && results[0][1] === "OK") {
+            success = true;
+            transactionError = false;
+            return transactionError;
```
Same here, don't set anything, just return!

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -47,6 +48,54 @@ export class WorkQueue {
+        } finally {
+          db.unwatch();
+        }
+      } while (transactionError);
+  
+      return success;
```
The loop can never be broken. Currently it always returns after setting `transactionError = false`. It's just a forever loop. No need for the return statement afterwards!

Also, there's no need for the `success` or `transactionError` variables. Just remove them.

You can write a loop that goes forever using
```javascript
for (;;) {
    // This will loop forever
}
```

Oxygen588 committed:
Necessary Fixes + Documentation.

Unnecessary code was removed.
Created returns for python function as well.
Fixed Length function error handling logics.
Created proper documentation and explanation of how this function should be used as well as what it does.

Oxygen588 committed:
Docs for Length function

Improved docs for the atomic lengths functions.

Oxygen588 committed:
Renamed functions from AddAtomicItem to AddNewItem as well as removed rust version.

Oxygen588 committed:
Delete redis-work-queue.sln

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -7,6 +7,11 @@ import Redis, {ChainableCommander} from 'ioredis';
+
+function delay(ms: number) {
+  return new Promise( resolve => setTimeout(resolve, ms) );
+}
```
I don't think this is used?

JOT85 commented on go/WorkQueue.go:
```diff
@@ -204,3 +271,12 @@ func (workQueue *WorkQueue) Complete(ctx context.Context, db *redis.Client, item
+		}
+	}
+	return false
+}
```
I don't think this is used?

JOT85 commented on rust/src/lib.rs:
```diff
@@ -277,9 +277,6 @@ impl WorkQueue {
 
-    /// Add an item to the work queue.
-    ///
-    /// This creates a pipeline and executes it on the database.
```
Why delete these docs?

JOT85 commented on tests/rust/src/main.rs:
```diff
@@ -24,6 +27,68 @@ fn main() -> RedisResult<()> {
+        items.push(item1.clone());
+        items.push(item2.clone());
+
+        //work_queue1.add_atomic_item(&mut client, &item1).await?;
+       // work_queue1.add_atomic_item(&mut client, &item2).await?;
+        work_queue1.add_atomic_item(client, &item1).await?;
+        work_queue1.add_atomic_item(client, &item2).await?;
```
The tests reference the function that was removed.

JOT85 commented on tests/rust/src/main.rs:
```diff
@@ -33,6 +98,11 @@ async fn async_main() -> RedisResult<()> {
+
+       if let Err(err) = run_duplicate_items_test().await {
+        eprintln!("Error: {:?}", err);
+    }
```
Format!

JOT85 commented on tests/run-test.sh:
```diff
@@ -105,4 +106,4 @@ if [[ "$tests" == *"node"* ]]; then
 fi
 
 echo "Running spawner..."
-python3 job-spawner-and-cleaner.py "$host" "$tests"
```
Did you intend to do this?

Oxygen588 committed:
Necessary fixes.

JOT85 committed:
Update docs, rename `Lengths` -> `Counts` and small tweaks to implementations:

- Go: return boolean from `WorkQueue.AddNewItem`

JOT85 committed:
Merge branch 'Atomic-length-method' of github.com:Oxygen588/redis-work-queue into Atomic-length-method

JOT85 commented on tests/rust/src/main.rs:
```diff
@@ -4,7 +4,10 @@ use futures_lite::future;
+
+
 use redis_work_queue::{Item, KeyPrefix, WorkQueue};
+use rand::Rng;
```
Rng is no longer used.

JOT85 commented on tests/run-test.sh:
```diff
@@ -42,6 +42,7 @@ while [[ $# -gt 0 ]]; do
   esac
 done
 
+python3 job-spawner-and-cleaner.py "$host" "$tests" &
```
Is this intentional?

JOT85 replied:
How come this moved from line 108 to line 45?

JOT85 commented on rust/src/lib.rs:
```diff
@@ -303,6 +303,20 @@ impl WorkQueue {
+        &'a self,
+        db: &mut C,
+    ) -> RedisResult<(u32,u32)> {
+        let (queue_length, processing_length): (u32, u32) = redis::pipe()
+            .atomic()
+            .llen(&self.main_queue_key)
+            .llen(&self.processing_key)
+            .query_async(db)
+            .await?;
+
+        Ok((queue_length, processing_length))
```
This doesn't need to be `async`? It can just return the future directly?

JOT85 commented on python/redis_work_queue/workqueue.py:
```diff
@@ -37,6 +37,37 @@ def add_item(self, db: Redis, item: Item) -> None:
+            except WatchError:
+                continue
+
+        pipeline.unwatch()
```
This is unreachable

JOT85 commented on python/redis_work_queue/workqueue.py:
```diff
@@ -37,6 +37,37 @@ def add_item(self, db: Redis, item: Item) -> None:
+            try:
+                pipeline.watch(self._main_queue_key, self._processing_key)
+
+                pipelineLPOS = db.pipeline()
```
Use snake case names in Python

JOT85 commented on python/redis_work_queue/workqueue.py:
```diff
@@ -37,6 +37,37 @@ def add_item(self, db: Redis, item: Item) -> None:
+        pipeline = db.pipeline(transaction=True)
+        while True:
+            try:
+                pipeline.watch(self._main_queue_key, self._processing_key)
```
Does this actually do the `WATCH` now, or does this only do the `WATCH` when the pipeline is executed? I don't know the answer.

Oxygen588 committed:
Fixes.

Python: Fixed Add_new_item, fixed exception handling.
Rust: Made `get_queue_lengths` not async, now returns future.
Test: There was no need for `python3 job-spawner-and-cleaner.py "$host" "$tests"` it was already running at line 45.

JOT85 commented on tests/rust/Cargo.toml:
```diff
@@ -11,3 +11,4 @@ futures-lite = "1"
 redis = "0.22"
 serde = "1"
 serde_json = "1"
+rand = "0.8.5"
```
I don't think this is used?

JOT85 commented on tests/rust/src/main.rs:
```diff
@@ -4,6 +4,8 @@ use futures_lite::future;
 use serde::{Deserialize, Serialize};
 
+
+
```
No need for the extra newlines!

JOT85 commented on tests/rust/src/main.rs:
```diff
@@ -24,6 +26,9 @@ fn main() -> RedisResult<()> {
 
+
+
+
```
No need for the extra newlines!

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -35,9 +35,9 @@
+            print(queue.get_queue_lengths(db))
 
-        print((shared_queue.queue_len(db), shared_queue.processing(db)))
+        print(shared_queue.get_queue_lengths(db))
```
I renamed this when writing docs and forgot to change the name in the tests!

JOT85 commented on tests/go/main.go:
```diff
@@ -46,6 +154,8 @@ func main() {
 	for {
 		shared = !shared
+		fmt.Println(goQueue.Lengths(ctx, db))
+		fmt.Println(sharedQueue.Lengths(ctx, db))
```
I don't think we should print these, it'll fill the logs up!

JOT85 replied:
And it's printed in the spawner already.

Oxygen588 committed:
Fixed addNewItem & tests for it.

Re-written the way tests are done for `addNewItem` now 3 functions that spawn 10 `addNewItem` calls for the item we are trying to add and runned concurrently.

TODO: Currently because of all added logic the difference in speed in between go and node & python, the difference in speed is much higher because of that the difference in number of shared-jobs done is much higher.

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,8 +121,117 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+	txf := func(tx *redis.Tx) error {
+
+		processingItemsInQueueCmd := tx.LPos(ctx, workQueue.processingKey, item.ID, redis.LPosArgs{
+			Rank:   0,
+			MaxLen: 0,
+		})
+		workingItemsInQueueCmd := tx.LPos(ctx, workQueue.mainQueueKey, item.ID, redis.LPosArgs{
+			Rank:   0,
+			MaxLen: 0,
+		})
+
+		_, ProcessingQueueCheck := processingItemsInQueueCmd.Result()
+		_, WorkingQueueCheck := workingItemsInQueueCmd.Result()
+
+		if ProcessingQueueCheck == redis.Nil && WorkingQueueCheck == redis.Nil {
+			time.Sleep(100 * time.Microsecond)
+			_, err := tx.TxPipelined(ctx, func(pipe redis.Pipeliner) error {
+				workQueue.AddItemToPipeline(ctx, pipe, item)
+				_, err := pipe.Exec(ctx)
+				return err
+			})
+			return err
+		} else {
+			return nil
+		}
+	}
+
+	for i := 0; i < 100; i++ {
+		err := db.Watch(ctx, txf, workQueue.processingKey, workQueue.mainQueueKey)
+		if err == nil {
+			return true, nil
+		}
+		if err == redis.TxFailedErr {
+			continue
+		}
+		return false, err
+	}
+
+	return false, errors.New("increment reached maximum number of retries")
+}
```
Could you please move this to the tests directory? You'll need to pass the keys as arguments, but it shouldn't be in the main library!

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,8 +121,117 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+		_, WorkingQueueCheck := workingItemsInQueueCmd.Result()
+
+		if ProcessingQueueCheck == redis.Nil && WorkingQueueCheck == redis.Nil {
+			time.Sleep(100 * time.Microsecond)
```
This definitely isn't long enough to garentee a collision in the tests!

Oxygen588 replied:
It meant to be Milliseconds!

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -61,7 +63,100 @@ export class WorkQueue {
+   * @param item The item that will be added, only if an item doesn't already exist with the same ID.
+   * @returns {boolean} returns false if already in queue or true if the item is successfully added.
+   */
+  async addNewItemWithSeep(db: Redis, item: Item): Promise<boolean> {
+    while (true) {
+      try {
+        await db.watch(this.mainQueueKey, this.processingKey);
+
+        const isItemInProcessingKey = await db.lpos(this.processingKey, item.id);
+        const isItemInMainQueueKey = await db.lpos(this.mainQueueKey, item.id);
+        if (isItemInProcessingKey !== null || isItemInMainQueueKey !== null) {
+          console.log("Item already exists, not added", item.id);
+          await db.unwatch();
+          return false;
+        }
+        await new Promise<void>((resolve) => setTimeout(resolve, 50));;
+        const transaction = db.multi();
+        this.addItemToPipeline(transaction, item);
+        const results = await transaction.exec();
+
+        if (!results) {
+          console.log("Transaction failed, item not added", item.id);
+          await db.unwatch();
+        }
+        console.log("Item added successfully", item.id);
+        return true
+      } catch (e) {
+        console.log("Error", e);
+      } finally {
+        await db.unwatch();
+      }
+    }
+    return false;
+  }
```
Could you please move this to the tests directory? You'll need to pass the keys as arguments, but it shouldn't be in the main library!

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -61,7 +63,100 @@ export class WorkQueue {
+          await db.unwatch();
+          return false;
+        }
+        await new Promise<void>((resolve) => setTimeout(resolve, 50));;
```
Bit longer sleep?

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -131,7 +242,7 @@ export class WorkQueue {
       maybeItemId = await db.rpoplpush(this.mainQueueKey, this.processingKey);
     }
-
+    console.log(`Leased ${maybeItemId}`)
```
Do we need that log?

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -207,7 +318,7 @@ export class WorkQueue {
       return false;
     }
-
+    console.log(`Completed ${item.id}`);
```
Do we need that log?

JOT85 commented on python/redis_work_queue/workqueue.py:
```diff
@@ -37,6 +36,41 @@ def add_item(self, db: Redis, item: Item) -> None:
+
+            except redis.WatchError:
+                continue
+            except Exception as e:
+                print("Error:", e)
+                raise
```
We don't need to print the error if we're raising it.

In fact, do we actually need this final `except`?

JOT85 commented on rust/src/lib.rs:
```diff
@@ -303,6 +303,22 @@ impl WorkQueue {
+        &'a self,
+        db: &'a mut C,
+    ) -> Box<dyn std::future::Future<Output = Result<(u32, u32), RedisError>> + 'a> {
+        Box::new(async move {
+            let result = redis::pipe()
+                .atomic()
+                .llen(&self.main_queue_key)
+                .llen(&self.processing_key)
+                .query_async(db)
+                .await?;
+
+            Ok(result)
```
Can we not just return

```rust
            redis::pipe()
                .atomic()
                .llen(&self.main_queue_key)
                .llen(&self.processing_key)
                .query_async(db)
```
?

Oxygen588 committed:
Fixes.

Moved `addNewItem` with sleep for tests, from the lib to the tests.

JOT85 committed:
rust: format and remove unused dependencies and imports

JOT85 committed:
go: AddNewItem: correct TxPipeline usage, correct error message, formatting

According to the docs at https://pkg.go.dev/github.com/redis/go-redis/v9#Client.Watch, we don't need
to `Exec` the pipeline.

Furthermore, the redis library ensures the database is safe to use multiple transactions
concurrently, so I've removed that from the doc comment.

JOT85 committed:
go: AddNewItem: remove retry limit

JOT85 committed:
go: Counts: format

JOT85 commented on python/redis_work_queue/workqueue.py:
```diff
@@ -1,10 +1,9 @@
-from redis import Redis
+import redis
 from redis.client import Pipeline
-
 from redis_work_queue import Item
 from redis_work_queue import KeyPrefix
-
+Redis = redis.Redis
```
@Oxygen588, how come?

Why not:
```python3
import redis
from redis import Redis
```
I don't know if there is a better option.

JOT85 replied:
Or even `from redis import Redis, WatchError`?

JOT85 committed:
python: format

JOT85 commented on go/WorkQueue.go:
```diff
@@ -120,8 +120,76 @@ func (workQueue *WorkQueue) AddItem(ctx context.Context, db *redis.Client, item
+
+		if ProcessingQueueCheck == redis.Nil && WorkingQueueCheck == redis.Nil {
+			_, err := tx.TxPipelined(ctx, func(pipe redis.Pipeliner) error {
+				workQueue.AddItemToPipeline(ctx, pipe, item)
+				return nil
+			})
+			return err
+		} else {
+			return nil
+		}
+	}
+
+	for {
+		err := db.Watch(ctx, txf, workQueue.processingKey, workQueue.mainQueueKey)
+		if err == nil {
+			return true, nil
+		}
```
This returns true even if the item isn't added!!

JOT85 committed:
node: AddNewItem: correct retry logic and remove logs

JOT85 commented on node/src/WorkQueue.ts:
```diff
@@ -61,7 +63,47 @@ export class WorkQueue {
+
+        const transaction = db.multi();
+        this.addItemToPipeline(transaction, item);
+        const results = await transaction.exec();
```
@Oxygen588 what does the return value look like in the success and failure case? I'm not sure what it would be, so it would be nice to have a comment!

JOT85 replied:
Sorry, to clarify what I mean, I don't know what the value of `results` if the transaction succeeds or fails.

JOT85 committed:
node: format

JOT85 committed:
rust: rename `get_queue_lengths` -> `counts` and add docs

Oxygen588 committed:
Small fixes

JOT85 commented on go/WorkQueue.go:
```diff
@@ -151,7 +151,7 @@ func (workQueue *WorkQueue) AddNewItem(
 			return err
 		} else {
-			return nil
+			return false
```
I'm not sure what's going on here? 😂

JOT85 committed:
Restore tests to upstream

JOT85 committed:
Readd tests/node/package-lock.json

JOT85 committed:
go: correctly return boolean indicating if an item was added in AddNewItem

Also format

JOT85 reviewed (approved):
Still needs tests, but we'll merge it into `atomic-methods` and add tests on that branch.
